go/vt/vtgate/planbuilder/select2.go

```diff
@@ -0,0 +1,367 @@
+// Copyright 2016, Google Inc. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package planbuilder
+
+import (
+	"encoding/json"
+	"errors"
+
+	"github.com/youtube/vitess/go/vt/sqlparser"
+)
+
+// PlanBuilder represents any object that's used to
+// build a plan. The top-level PlanBuilder will be a
+// tree that points to other PlanBuilder objects.
+// Currently, JoinBuilder and RouteBuilder are the
+// only two supported PlanBuilder objects. More will be
+// added as we extend the functionality.
+// Each Builder object builds a Plan object, and they
+// will mirror the same tree. Once all the plans are built,
+// the builder objects will be discarded, and only
+// the Plan objects will remain.
+type PlanBuilder interface {
+	// Order is a number that signifies execution order.
+	// A lower Order number Route is executed before a
+	// higher one. For a node that contains other nodes,
+	// the Order represents the highest order of the leaf
+	// nodes.
+	Order() int
```

Comment:
For now, I plan to use it for deciding where to push a WHERE clause. The highest order == rightmost chunk. But the order could also be used for other things. At this point, I don't know where else it will be used.